adage/analyze/management/commands/import_gene_sample_expr.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python
+
+"""
+This management command reads an input file of gene-sample expression values
+and loads the valid data into database.  It should be invoked like this:
+
+  python manage.py import_gene_sample_expr <expression_filename>
+
+If a data source (on the first row) or gene name (on the first column) is
+not found in the database, a warning message will be generated and the
+corresponding column or row will be skipped.
+"""
+
+
+from __future__ import print_function
+from django.core.management.base import BaseCommand, CommandError
+from django.db import transaction
+from genes.models import Gene
+from analyze.models import Sample, ExpressionValue
+
+import logging
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+
+class Command(BaseCommand):
+    help = ("Import gene-sample expression values from an input file.")
+
+    def add_arguments(self, parser):
+        parser.add_argument('expression_filename', type=file)
+
+    def handle(self, **options):
+        try:
+            import_expr(options['expression_filename'])
+            self.stdout.write(self.style.NOTICE(
+                "Imported gene-sample expression values successfully"))
+        except Exception as e:
+            raise CommandError(
+                "Raised exception when importing gene-sample expression "
+                "values: %s" % e)
+
+
+def import_expr(file_handle):
+    """
+    Function that reads input file and load gene-sample expression values
+    into the database.
+    """
+
+    # Enclose reading/importing process in a transaction context manager.
+    # Any exception raised inside the manager will terminate the transaction
+    # and roll back the database.
+    with transaction.atomic():
+        samples = []
+        for line_index, line in enumerate(file_handle):
+            tokens = line.rstrip('\r\n').split('\t')
+            if line_index == 0:
+                tokens = tokens[1:]
+                read_header(tokens, samples)
+            else:
+                import_data_line(line_index + 1, tokens, samples)
+
+
+def read_header(tokens, samples):
+    """
+    Read input tokens on header line and save the corresponding sample
+    object into "samples". (Each token will be searched in the database
+    using "ml_data_source" field. If a token does not match any sample's
+    ml_data_source, put None into "samples".)
+
+    An exception will be raised if any of the following errors is detected:
```

Comment:
The plural form is correct.

For such grammatical "edge cases," I usually turn to *The Elements of Style*, by William Strunk, Jr., and E.B. White, more commonly known as "Strunk & White." The relevant rule is described in Strunk & White in Section I, part 9: "The number of the subject determines the number of the verb." In this case, the part that's tricky is that the subject of the sentence is "any of."

The example given in the book is:

> A common blunder is the use of a singular verb form in a relative clause following "one of ..." or a similar expression when the relative is the subject.

and then two examples are given showing the wrong wording in a column on the left and the corrected wording in a column on the right. Rather than type them all out, the corrected first example says

> One of the ablest men who have attacked this problem

But we are not writing a thesis here, so I'm good with it either way. Just thought you'd like to know about Strunk & White. It is considered an authoritative source and a worthwhile companion for any writer of English.